### PR TITLE
Avoid `rustup` invocation for non-rustup rust installation

### LINKED ISSUE #4218: Avoid calling `rustup` in tests
I'm on [NixOS](https://nixos.org/nixos/) and use [nixpkgs-mozilla](https://github.com/mozilla/nixpkgs-mozilla) to setup `cargo`, `rustc` and other components, **without `rustup`**.

 `cargo test` in rust-analyzer results in 3 failure for me:
- cli::check_code_formatting
- cli::generated_assists_are_fresh
- cli::generated_grammar_is_fresh

They are all trying to invoke `rustup run stable -- ...` and  then `rustup toolchain install stable`. Obviously, they failed because there is no `rustup`.
But I already have `cargo` and `rustfmt` installed!

Why not invoking `cargo ...` or `rustfmt ...` directly? Since we are expecting `stable`, I think it should not lead to problem even if nightly components are invoked.

### PR DESCRIPTION
**Pull request by oxalica**
Fix #4218 and #3243.

### CODE REVIEW

File: xtask/src/lib.rs [hunk 1]
```diff
@@ -79,29 +88,42 @@ fn reformat(text: impl std::fmt::Display) -> Result<String> {
...
     {
-        Ok(status) if status.success() => return Ok(()),
-        _ => (),
-    };
-    run!("rustup toolchain install {}", TOOLCHAIN)?;
-    run!("rustup component add rustfmt --toolchain {}", TOOLCHAIN)?;
-    Ok(())
+        Ok(output)
+            if output.status.success()
+                && std::str::from_utf8(&output.stdout)?.contains(RUSTFMT_TOOLCHAIN) =>
```
Comment (bjorn3) [inline xtask/src/lib.rs:101]: This doesn't check that it is the latest version. `rustup toolchain install` updates if the toolchain isn't the latest version.

Comment (oxalica) [reply xtask/src/lib.rs:101]: Does outdated clippy really matter? I also think it's developer's responsibility to choose the correct version of clippy.

Note that the aim of this PR is to make tests and utilities work with bare-installation of rust with components, **without rustup**.

**Comment by matklad** [comment]
bors r+

yup, that's a much better approach to doing this. After this PR, I'd like to tweak `run!` macro to maybe handle env as well, but I'll do this myself as I am not yet sure how it should look.

**Comment by polyzen** [comment]
Still getting the test failures with rust-analyzer 2020-05-04 and rustfmt provided by the Arch Linux rust package:
```
> rustfmt --version
rustfmt 1.4.12-
thread 'cli::check_code_formatting' panicked at 'Failed to run rustfmt from toolchain 'stable'. ...
```

For
- cli::check_code_formatting
- cli::generated_assists_are_fresh
- cli::generated_grammar_is_fresh

**Pull request merged**